ols/src/llms/providers/azure_openai.py

```diff
@@ -52,6 +54,47 @@ def default_params(self) -> dict[str, Any]:
             "verbose": False,
         }
 
+        if self.credentials is not None:
+            # if credentials with API key is set, use it to call Azure OpenAI endpoints
+            default_parameters["api_key"] = self.credentials
+        else:
+            # credentials for API key is not set
+            # -> it is needed to retrieve token first
+            access_token = self.retrieve_access_token(azure_config)
+            default_parameters["azure_ad"] = access_token.token
+
+        return default_parameters
+
     def load(self) -> LLM:
         """Load LLM."""
         return AzureChatOpenAI(**self.params)
+
+    def retrieve_access_token(self, azure_config: AzureOpenAIConfig) -> AccessToken:
+        """Retrieve access token to call Azure OpenAI."""
+        if azure_config is None:
+            raise ValueError(
+                "Credentials for API token is not set and "
+                "Azure-specific parameters are not provided."
+                "It is not possible to retrieve access token."
+            )
+        if azure_config.tenant_id is None:
+            raise_missing_attribute_error("tenant_id")
+        if azure_config.client_id is None:
+            raise_missing_attribute_error("client_id")
+        if azure_config.client_secret is None:
+            raise_missing_attribute_error("client_secret")
+
+        # everything is there, try to retrieve credential
+        credential = ClientSecretCredential(
+            azure_config.tenant_id, azure_config.client_id, azure_config.client_secret
+        )
+        return credential.get_token("https://cognitiveservices.azure.com/.default")
```

Comment:
Could it fail? Should we catch an exception in that case?

Reply:
it can fail and will fail, is it a problem?

Reply:
Maybe logging something like the service is not available instead of hard crash?

Reply:
ok will change it. It will fail later, but that's another story

Reply:
updated @onmete